### ISSUE TITLE
YOLO

### ISSUE THREAD

**Issue by iwasakishuto** [iwasakishuto/Papers-With-Comments#2]
# YOLO
- You Only Look Once: Unified, Real-Time Object Detection
- Submitted on Mon, 8 Jun 2015 (v1)

### Link
- [paper](https://arxiv.org/pdf/1506.02640.pdf)
- [Github](https://github.com/pjreddie/darknet/wiki/YOLO:-Real-Time-Object-Detection)
- [Project Webpage](https://pjreddie.com/darknet/yolo/)

### Summary
- Frame object detection as a **"regression problem"** to spatially separated bounding boxes and associated class probabilities. This enable YOLO to train on loss function **directly corresponds to entire detection performance.**
- Since the whole detection pipeline is a **single network**, it **can be optimized end-to-end directly on detection performance**, and **extremely fast**.
- makes **more localization errors** but is **less likely to predict false positives on background.**

#### Detection Pipeline
**※ Apply a single neural network.**

1. Divides the input image into an "S × S" regions.
2. **Each** grid cell predicts "B" **bounding boxes** and **Confidence scores** for the boxes.
    - Formally they define confidence as "Pr(Object)*IOU(Area of Overlap/Area of Union)".
    - If no object exists in the cell, the confidence scores should be zero.
    - Each **bounding box** consists of 5 predictions: x,y,w,h,and confidence.
    - Each **grid cell** also predicts "C" conditional class probabilities "Pr(Class_i | Object)".
3. At test time, we multiply the conditional class probabilities and the individual box confidence predictions:
Pr(Class_i | Object) * Pr(Object) * IOU = Pr(Class_i) * IOU

![Figure 2](https://miro.medium.com/max/2304/1*m8p5lhWdFDdapEFa2zUtIA.jpeg)

#### Training
- Pretrain on the [ImageNet 1000-class competition dataset](https://arxiv.org/abs/1409.0575).
- Use the **[Darknet framework](https://pjreddie.com/darknet/)** for all training and inference.
- Add four convolutional layers and two fully connected layers **for the purpose of improving performance**: ["Object detection networks on convolutional feature maps."](https://arxiv.org/abs/1504.06066)
- As detection often requires fine-grained visual information, **increase the input resolution from 224×224 to 448×448.**
- Normalize the image width and height to [0,1].
- Parametrize the bounding box x and y coordinates to be **offsets of the particular grid cell location** so they are also bounded between 0 and 1.
- **Linear** activation function for the final layer and **Leaky rectified linear** (`φ(x) = x if x>0 else 0.1*x`) activation for the others.
- Optimize for **sum-squared error** in the output because it is easy to optimize. **(However it does not perfectly align with goal of maximizing average precision.)**
- In every image **many grid cells do not contain any object**, so preventing overpowering the gradient from other cells, **weight depends on whether the cell contains objects or not.**
- Predict the square root of the bounding box width and height so that **it reflect that small deviations in large boxes matter less than in small boxes.**
- In the training time, they assign one predictor to be **"responsible for"** predicting an object **based on which prediction has the highest current IOU** with the ground truth.
- The loss function composes of:
    - the **localization loss** (errors between the predicted boundary box and the ground truth) → **the 1st and 2nd terms**
    - the **classification loss.** → **the 3rd and 4th terms**
    - the **confidence loss** (the objectness of the box). → **the 5th term**

![YOLO Loss function](https://miro.medium.com/max/2744/1*aW6htqx4Q7APLrSQg2eWDw.png)

### Limitation
- Since each grid cell only predicts few boxes and can only have one class, **YOLO struggles with small objects that appear in groups**, such as flocks of birds.
- Since YOLO learns to predict bounding boxes from data, it struggles to generalize to **objects in new or unusual aspect ratios or configurations.**
- Loss function **treats error the same in "small bounding boxes" VS "large bounding boxes."**

### Comparison

|Name|Explanation|
|:-:|:-|
|Deformable parts model **(DPM)**|Use a sliding window approach to object detection and **its pipeline is disjoint.**|
|R-CNN|Use region proposals to find objects in images and its pipeline is very complicated ([Selective Search](http://www.huppelen.nl/publications/selectiveSearchDraft.pdf) generates potential bounding boxes, a convolutional network extracts features, an SVM scores the boxes, a linear model adjust the bounding boxes.) and must be **precisely tuned independently.**|
|Faster R-CNN|Focus on speeding up the R-CNN framework but still fall short of real-time performance.|
|[Deep MultiBox](https://arxiv.org/pdf/1312.2249.pdf)|Train a convolutional network to predict regions of interest instead of using Selective Search.|
|[OverFeat](https://arxiv.org/pdf/1312.6229.pdf)|Train a convolutional neural network to perform localization and adapt that localizer to perform detection.|
|[MultiGrasp](https://arxiv.org/pdf/1412.3128.pdf)|YOLO's grid approach to bounding box prediction is based on this system for regression to grasps.|
